test/IISIntegration.FunctionalTests/Utilities/IISApplication.cs

```diff
@@ -0,0 +1,173 @@
+            {
+                if ((string)element["name"] == @"AspNetCoreModuleV2")
+                {
+                    _originalAspNetCoreDllLocation = (string)element["image"];
+                    element["image"] = ancmFile;
+                    found = true;
+                    break;
+                }
+            }
+
+            // If ANCMV2 isn't in the global apphost.config, add it there.
+            if (!found)
+            {
+                _originalAspNetCoreDllLocation = "";
+                var addElement = globalModulesSection.GetCollection().CreateElement("add");
+                addElement["name"] = "AspNetCoreModuleV2";
+                addElement["image"] = ancmFile;
+                globalModulesSection.GetCollection().Add(addElement);
+            }
+        }
+
+        private string GetAncmLocation(string dllRoot)
+        {
+            var arch = _deploymentParameters.RuntimeArchitecture == RuntimeArchitecture.x64 ? @"x64\aspnetcorev2.dll" : @"x86\aspnetcorev2.dll";
+            var ancmFile = Path.Combine(dllRoot, arch);
+            if (!File.Exists(Environment.ExpandEnvironmentVariables(ancmFile)))
+            {
+                ancmFile = Path.Combine(dllRoot, "aspnetcorev2.dll");
+                if (!File.Exists(Environment.ExpandEnvironmentVariables(ancmFile)))
+                {
+                    throw new FileNotFoundException("AspNetCoreModule could not be found.", ancmFile);
+                }
+            }
+
+            return ancmFile;
+        }
+
+        public void StopAndDeleteAppPool()
+        {
+            if (string.IsNullOrEmpty(WebSiteName))
+            {
+                return;
+            }
+
+            RemoveSite();
+
+            RemoveAppPool();
+
+            RemoveAppHostConfig();
+        }
+
+        private void RemoveAppPool()
+        {
+            var appPool = _serverManager.ApplicationPools.FirstOrDefault(pool => pool.Name == AppPoolName);
+            if (appPool != null)
+            {
+                appPool.Stop();
+                _serverManager.ApplicationPools.Remove(appPool);
+                _serverManager.CommitChanges();
+            }
+        }
+
+        private void RemoveSite()
+        {
+            var siteToRemove = _serverManager.Sites.FirstOrDefault(site => site.Name == WebSiteName);
+            if (siteToRemove != null)
+            {
+                siteToRemove.Stop();
+                _serverManager.Sites.Remove(siteToRemove);
+                _serverManager.CommitChanges();
+            }
+        }
+
+        private void RemoveAppHostConfig()
+        {
+            var config = _serverManager.GetApplicationHostConfiguration();
+            var globalModulesSection = config.GetSection("system.webServer/globalModules");
+            foreach (var element in globalModulesSection.GetCollection())
+            {
+                if ((string)element["name"] == @"AspNetCoreModuleV2")
```

Comment:
What about the case where AspNetCoreModuleV2 had been added? Should it be removed again?

Reply:
I think we need to carry our own apphostconfig backup old one and use ours for tests to avoid machine state affecting tests.

Reply:
That's dramatic. Even more reason to work out a way to prevent parallel execution across repos.